### PR TITLE
Feat/radius op

### PR DESCRIPTION
**Pull request by yaoyic**
# PR Checklist

- [ ] Bug fix
- [x] Feature addition/change
- [ ] Documentation addition/change
- [ ] Test addition/change
- [ ] Black formatting

---

### Describe your changes here:
Paul Mifsud wrote a nice CUDA kernel for finding the spatial neighbors for graph neural networks. He also mentioned that there is a further speed up with merged distance computation, but the distance output does not have a backward pass to the input coordinates.
I made the tests more concise, integrated it to the schnet and wrote a `setup.py` for installing it as an independent optional package. It will be automatically used for CUDA platform if installed in your python environment.
The remaining TODOs are to have Paul's backward pass implementation included, perform a benchmark, make nicer tests, and maybe incorporate the package better to the main code base.
Note: the implementation so far does not support periodic boundary conditions.

### CODE REVIEW

File: mlcg/nn/schnet.py [hunk 1]
```diff
@@ -267,7 +295,7 @@ def __init__(
         self.cutoff = cutoff
         self.reset_parameters()
 
-    def reset_parameters(self):
+    aef reset_parameters(self):
```
Comment (nec4) [inline mlcg/nn/schnet.py:298]: I think the tests are primarily failing because of this.

**Comment by yaoyic** [comment]
@paulmifsud @sayeg84 The code can be merged now. 
There are still problems to solve, such as a more consistent installation, and more efficient backward implementation. But since they don't prevent us from using the functionality as is, nor do they affect the compatibility, I think we can deal with them in another PR.
I have ensured that the code can run Nick's models with errors up to numerical differences (energy and force ATOL<2e-4 for float32 types, I think it's related to the order of summation, etc).
On the speed benchmarks: the speed is comparable with the old version with `torch_cluster`. Training of schnet model at transferable resolution with batch CATH:DIMER=16:4 results in 3.24 batches / s on an RTX 3090 (power capped at 310W) for both the original and the new implementation.

**Comment by paulmifsud** [comment]
@yaoyic @sayeg84 wrt to build options what are the needs? and why do neither of JIT nor setup.py satisfy the need? wrt numerical discrepancies between different kernels, that's just floating point arithmetic but one can look into and improve accuracy if needed, wrt to speedup, the provided kernel achieves a speedup of 1.5x on A100 (on consumer grade GPUs it performs worse due to different SM cache and latency between cache and global memory) on graphs of the order of 100000, with CUDA kernels you always have to tune them to specific expected input sizes (and hardware) as they will perform differently based on such requirements, more work is planned in this direction, wrt to backward logic, I have not wrote any CUDA for it yet because first it makes sense to tune the kernel to a specific size, then do profiling, then include any backwards logic that merits it, also the index_add() is a similiar problem to scatter_add() which I am still working on

**Review by sayeg84 — APPROVED** [review]
@yaoyic @paulmifsud as Yaoyi says, I think we can go for this merge now. 

I've noticed some of the problems with numerical accuarcy before (i.e. differences in ATOL 5e-4 in energy eval for the same graph) but I don't think there will be a way to solve them.

**Comment by yaoyic** [comment]
Thanks a lot @paulmifsud for your follow up!
> @yaoyic @sayeg84 wrt to build options what are the needs? and why do neither of JIT nor setup.py satisfy the need? 

Currently in the code we are doing both. I mean we might merge it to one strategy one day. The complain about JIT is that on computational nodes of some of our clusters the cache might not be saved and accessible by others, thus it is compiled repetitively. With `setup.py`, if we don't hard code every possible compute capabilities (leading to long compile times), one needs to manually set the cuda compute versions with `TORCH_CUDA_ARCH_LIST` for every GPU type one wants to run it on. That's not friendly for normal users.

> wrt numerical discrepancies between different kernels, that's just floating point arithmetic but one can look into and improve accuracy if needed, 

As I said I think it's merely due to the order of summation, and I am not quite sure about how to solve that on GPUs (on CPU I will use e.g., [Kahan](https://en.wikipedia.org/wiki/Kahan_summation_algorithm)). Since the other method we have been using previously or in your test might just be using a different summing order, I wouldn't see it as a problem.

> wrt to speedup, the provided kernel achieves a speedup of 1.5x on A100 (on consumer grade GPUs it performs worse due to different SM cache and latency between cache and global memory) on graphs of the order of 100000, with CUDA kernels you always have to tune them to specific expected input sizes (and hardware) as they will perform differently based on such requirements, more work is planned in this direction, 

Thanks for the benchmark. If you mean the number of graph nodes in "graphs of the order of 100000", then sadly for mlcg we are not dealing with anything close to that (I think practically we are doing graphs with between 10 and several thousand nodes). For larger systems I don't think SchNet or any other net we have in mind can handle them efficiently.

> wrt to backward logic, I have not wrote any CUDA for it yet because first it makes sense to tune the kernel to a specific size, then do profiling, then include any backwards logic that merits it, also the index_add() is a similiar problem to scatter_add() which I am still working on

Then let's leave it as is for now?

**Comment by paulmifsud** [comment]
@yaoyic 

wrt to build options, I thought JIT made more sense because in practice, it will take maybe 2 minutes to compile which is insignificant when the job is going to run for multiple days

wrt to numerical discrepancies, there are options to reduce the accumulation of numerical errors, maybe we should discuss the necessity of such measures

wrt to speed up, it depends also how big your batch/gpu is, but I believe I can tune the kernel to be more performant at the desired range

**Comment by sayeg84** [comment]
@yaoyic @yaoyic expanding a bit on the numerical accuarcy: I would think its relevant that we make it as small as possible. I would think that maybe its not due only to the order of summation but maybe a limitation of the scatter sum (see https://github.com/rusty1s/pytorch_scatter/issues/450, https://github.com/rusty1s/pytorch_scatter/issues/163).
